### PR TITLE
Add LiveUpdater and LiveUpdateHandler interface

### PR DESCRIPTION
**Pull request by acardace**
### What this PR does
This PR adds the `LiveUpdater` which is used for registering types implementing the LiveUpdateHandler interface.

Types implementing `LiveUpdateHandler` are used to manage a subset of live-updatable fields in the VMSpec, they just have to specify which fields they manage and a liveUpdateHandler implementation to react to changes to those fields.

This will help clean up the current implementation of the vmRolloutStrategy feature and should make it easier to introduce new live-updatable fields.

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
NONE
```

### CODE REVIEW

**Comment by acardace** [comment]
@jean-edouard @xpivarc if you want to start taking a look at the approach and have some early feedback that'd be great.

File: pkg/virt-controller/liveupdate/memory.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+		newMemoryReq = *vm.Spec.Template.Spec.Domain.Memory.Guest
+		// adjusting memoryDelta too for the new limits computation (if required)
+		memoryDelta = resource.NewQuantity(vm.Spec.Template.Spec.Domain.Memory.Guest.Value()-newMemoryReq.Value(), resource.BinarySI)
+	}
+
+	patches := []string{
+		fmt.Sprintf(`{ "op": "test", "path": "/spec/domain/memory/guest", "value": "%s"}`, vmi.Spec.Domain.Memory.Guest.String()),
+		fmt.Sprintf(`{ "op": "replace", "path": "/spec/domain/memory/guest", "value": "%s"}`, vm.Spec.Template.Spec.Domain.Memory.Guest.String()),
+		fmt.Sprintf(`{ "op": "test", "path": "/spec/domain/resources/requests/memory", "value": "%s"}`, vmi.Spec.Domain.Resources.Requests.Memory().String()),
+		fmt.Sprintf(`{ "op": "replace", "path": "/spec/domain/resources/requests/memory", "value": "%s"}`, newMemoryReq.String()),
```
Comment (alicefr) [inline pkg/virt-controller/liveupdate/memory.go:124]: Please use the new patchSet type

**Comment by alicefr** [comment]
@acardace I like the idea!
it will be god to see also which parts of the code are replaced by this new package, but I understand that this still in draft

File: pkg/virt-controller/liveupdate/handlers.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+				return fmt.Errorf("cannot register live-update handler, field '%s' does not exist", field)
+			}
+		}
+
+		up.handlers = append(up.handlers, handler)
+	}
+	return nil
+}
+
+func (up *LiveUpdater) FilterUpdatableFields(vm, oldVM *v1.VirtualMachine) error {
```
Comment (alicefr) [inline pkg/virt-controller/liveupdate/handlers.go:121]: Would be possible to overwrite this function per handler too? In the case of volume update we might want to distinguish per volume type (see in https://github.com/kubevirt/kubevirt/pull/11700 the function [liveUpdateVolumes](https://github.com/alicefr/kubevirt/blob/a4b6a60026148f7f5e6289769e721491c39dc245/pkg/virt-controller/watch/vm.go#L2932))

File: pkg/virt-controller/liveupdate/handlers.go [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+		for _, field := range handler.GetManagedFields() {
+			oldValue, err := getField(oldVM, field)
+			if err != nil {
+				return err
+			}
+
+			newValue, err := getField(vm, field)
+			if err != nil {
+				return err
+			}
```
Comment (alicefr) [inline pkg/virt-controller/liveupdate/handlers.go:141]: Actually, the same comment https://github.com/kubevirt/kubevirt/pull/11748/files#r1572337056 is true here. For the volumes we might want to filter certain volumes types. An the getField for new and old values isn't the same.

Comment (alicefr) [reply pkg/virt-controller/liveupdate/handlers.go:141]: We would need to cover cases like the one described in https://github.com/kubevirt/kubevirt/pull/11700#discussion_r1572173667 and this requires the old and new version of the VM.

File: pkg/virt-controller/liveupdate/handlers.go (hunk 2 above)
Comment (acardace) [reply pkg/virt-controller/liveupdate/handlers.go:121]: Nope, I don't think it's possible, I might have to get rid of reflection to support that use case.

File: pkg/virt-controller/liveupdate/memory.go (hunk 1 above)
Comment (acardace) [reply pkg/virt-controller/liveupdate/memory.go:124]: I didn't because it's already done here https://github.com/kubevirt/kubevirt/pull/11655.

**Comment by acardace** [comment]
> @acardace how do you plan to update the VirtualMachineRestartRequired condition if there are multiple reasons why it should be updated. E.g update cpu AND memory. Would you run all the handlers and concatenate the reasons?

I don't plan to, only 1 handler will be run at a time. This is actually a bug in the current implementation (the design stated this should not be done IIRC).

File: pkg/virt-controller/liveupdate/handlers.go (hunk 3 above)
Comment (acardace) [reply pkg/virt-controller/liveupdate/handlers.go:141]: A version without reflection that allows the type implementing the interface to do this was already in the works.

**Review by enp0s3 — COMMENTED** [review]
@acardace Hey! do you have any reference to a discussion or a design grooming about this feature?

**Pull request closed without merging**